### ISSUE TITLE
since 4.19, workspaces don't move to configured output when it gets available

### ISSUE THREAD

**Issue by Hippo0o** [i3/i3#4261]
## I'm submitting a…

<pre>
[x] Bug
[ ] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

##### example config: `workspace 1 output DisplayPort-0`
When `DisplayPort-0` is not connected, `workspace 1` displays on a different output.
When `DisplayPort-0` gets connected, `workspace 1` doesn't move to output `DisplayPort-0` like it did pre 4.19.

## Expected Behavior

When the configured output gets available, the workspace should move to the output like in pre 4.19 versions.

## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version:  4.19
</pre>

<pre>
- Linux Distribution & Version: Arch Linux
- Are you using a compositor (e.g., xcompmgr or compton): xcompmgr
</pre>

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow https://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by orestisfl** [comment]
Can't test right now but maybe something with https://github.com/i3/i3/commit/0d5a7eeff8e95277256fb191f80112296e0aba6f or (less likely) https://github.com/i3/i3/commit/831a52de9abae1837077ef544e9c3ce52b5d2848

**Comment by orestisfl** [comment]
Can you please paste the full config you used or the complete log file like the bot requests?

**Comment by Hippo0o** [comment]
I will do it tomorrow.

**Comment by orestisfl** [comment]
Oh and if it is not too much work a log with 4.18 and the same situation will be useful.

**Comment by Hippo0o** [comment]
@orestisfl scenario: 
- start i3
- open termite on workspace f1 on output DVI-D-0 and workspace 1&2 on output DisplayPort-0
- workspace 2 focused on output DisplayPort-0
- switch off output DisplayPort-0 via mode $mode_display:
  - `bindsym h exec --no-startup-id $HOME/.screenlayout/hostmode.sh` *(xrandr --output DVI-D-0 --primary --mode 1920x1080 --pos 0x0 --rotate normal --output DisplayPort-0 --off --output HDMI-A-0 --off)*
  - `bindsym Escape mode "default", restart`
- switch on output DisplayPort-0 via mode $mode_display:
  - `bindsym d exec --no-startup-id $HOME/.screenlayout/monitor.sh` *(xrandr --output DVI-D-0 --mode 1920x1080 --primary --pos 0x0 --rotate normal --output HDMI-A-0 --off --output DisplayPort-0 --mode 1920x1080 --rate 240 --pos 1920x0 --rotate normal)*
  - `bindsym Escape mode "default", restart`

see config in logs.

https://logs.i3wm.org/logs/5632642737242112.bz2 4.18.3
https://logs.i3wm.org/logs/5769834260332544.bz2 4.19

**Comment by orestisfl** [comment]
Haven't got to the bottom of this and I am not able to reproduce it yet.
From the logs it doesn't seem to be https://github.com/i3/i3/commit/0d5a7eeff8e95277256fb191f80112296e0aba6f. It doesn't seem to be anything in randr.c in general just from diffing the logs:

## first

```
randr.c:randr_query_outputs_15 - Querying outputs using RandR 1.5
randr.c:randr_query_outputs_15 - 2 RandR monitors found (timestamp 31493357)
randr.c:randr_query_outputs_15 - name DVI-D-0, x 0, y 0, width 1920 px, height 1080 px, width 598 mm, height 336 mm, primary 1, automatic 1
randr.c:randr_query_outputs_15 - name DisplayPort-0, x 1920, y 0, width 1920 px, height 1080 px, width 598 mm, height 336 mm, primary 0, automatic 1
randr.c:randr_query_outputs - Active RandR output found. Disabling root output.
randr.c:randr_query_outputs - output 0x55aea6461d50 / DVI-D-0, position (0, 0), checking for clones
randr.c:randr_query_outputs - output 0x55aea6414cc0 / DisplayPort-0, position (1920, 0), checking for clones
randr.c:randr_query_outputs - Need to initialize a Con for output DVI-D-0
randr.c:output_init_con - init_con for output DVI-D-0
randr.c:output_init_con - Using existing con 0x55aea64228a0 / DVI-D-0
randr.c:output_init_con - Not adding workspace, this was a reused con
randr.c:randr_query_outputs - Need to initialize a Con for output DisplayPort-0
randr.c:output_init_con - init_con for output DisplayPort-0
randr.c:output_init_con - Using existing con 0x55aea6427280 / DisplayPort-0
randr.c:output_init_con - Not adding workspace, this was a reused con
randr.c:randr_query_outputs - Focusing primary output DVI-D-0
randr.c:get_output_containing - comparing x=1919 y=552 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1920 y=1055 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1920 y=1055 with x=1920 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=0 y=1055 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1920 y=552 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1920 y=552 with x=1920 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1920 y=552 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1920 y=552 with x=1920 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=2880 y=552 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=2880 y=552 with x=1920 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1921 y=552 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1921 y=552 with x=1920 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1922 y=551 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1922 y=551 with x=1920 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1923 y=550 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1923 y=550 with x=1920 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1924 y=550 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1924 y=550 with x=1920 and y=0 width 1920 height 1080
```

## second

```
randr.c:randr_query_outputs_15 - Querying outputs using RandR 1.5
randr.c:randr_query_outputs_15 - 2 RandR monitors found (timestamp 31714768)
randr.c:randr_query_outputs_15 - name DVI-D-0, x 0, y 0, width 1920 px, height 1080 px, width 598 mm, height 336 mm, primary 1, automatic 1
randr.c:randr_query_outputs_15 - name DisplayPort-0, x 1920, y 0, width 1920 px, height 1080 px, width 598 mm, height 336 mm, primary 0, automatic 1
randr.c:randr_query_outputs - Active RandR output found. Disabling root output.
randr.c:randr_query_outputs - output 0x55d1981cde90 / DVI-D-0, position (0, 0), checking for clones
randr.c:randr_query_outputs - output 0x55d1981df6c0 / DisplayPort-0, position (1920, 0), checking for clones
randr.c:randr_query_outputs - Need to initialize a Con for output DVI-D-0
randr.c:output_init_con - init_con for output DVI-D-0
randr.c:output_init_con - Using existing con 0x55d19819a340 / DVI-D-0
randr.c:output_init_con - Not adding workspace, this was a reused con
randr.c:randr_query_outputs - Need to initialize a Con for output DisplayPort-0
randr.c:output_init_con - init_con for output DisplayPort-0
randr.c:output_init_con - Using existing con 0x55d1981a0800 / DisplayPort-0
randr.c:output_init_con - Not adding workspace, this was a reused con
randr.c:randr_query_outputs - Focusing primary output DVI-D-0
randr.c:get_output_containing - comparing x=1919 y=552 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1920 y=1055 with x=0 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=1920 y=1055 with x=1920 and y=0 width 1920 height 1080
randr.c:get_output_containing - comparing x=0 y=1055 with x=0 and y=0 width 1920 height 1080
```

**Comment by Hippo0o** [comment]
> switch on output DisplayPort-0 via mode $mode_display:
bindsym d exec --no-startup-id $HOME/.screenlayout/monitor.sh (xrandr --output DVI-D-0 --mode 1920x1080 --primary --pos 0x0 --rotate normal --output HDMI-A-0 --off --output DisplayPort-0 --mode 1920x1080 --rate 240 --pos 1920x0 --rotate normal)
bindsym Escape mode "default", restart

Might be important to note that only the first workspace in order is moved to output DisplayPort-0.
When workspace 1&2 available, only workspace 1 will go to output DisplayPort-0.
When workspace 3&5 available, only workspace 3 will go to output DisplayPort-0.

**Comment by lukrop** [comment]
I'm experiencing the same. Workspaces won't move to newly connected outputs according to config as it did in 4.18. Please let me know which logs or infos I could provide to help out.

**Comment by alienth** [comment]
Also experiencing this since upgrading from 4.18 to 4.19. I can very reliably reproduce by turning off one display and turning it back on. Likewise let me know if I can provide any logs or perform any debugging steps.

**Comment by bw0248** [comment]
Yes I'm having the same issue since upgrading to 4.19. All workspaces stay on the previously connected monitor except for the active workspace. New workspaces open on the correct monitor though.

**Comment by Zeioth** [comment]
As stated in #4278 I can confirm this bug too.
Also, I can't revert to 1.18 on arch using 'downgrade'.

**Comment by matejdro** [comment]
Workaround. I put this on the end of my monitor enabling script:

```
i3-msg -t run_command 'workspace "5"; move workspace to output DisplayPort-1;'
i3-msg -t run_command 'workspace "6"; move workspace to output DisplayPort-1;'
i3-msg -t run_command 'workspace "7"; move workspace to output DisplayPort-1;'
i3-msg -t run_command 'workspace "8"; move workspace to output DisplayPort-1;'
i3-msg -t run_command 'workspace "9"; move workspace to output DisplayPort-1;'
i3-msg -t run_command 'workspace "10"; move workspace to output DisplayPort-1;'
```

(of course, edit for your own workspaces and display ports)

**Comment by abeutot** [comment]
Hello.

Having the same issue I did a bisect and the issue is caused by 0d5a7eeff8e95277256fb191f80112296e0aba6f.

In this commit, instead of iterating over assignation rules, we iterate over the workspaces directly. The issue seem to be caused by the fact that since we may move the workspace into another output (by calling the `workspace_move_to_output` function), we modify the list we are iterating on.
The example scenario works as follow:
`init_ws_for_output` is called with "OUTPUT-2"
We iterate over the workspaces of "OUTPUT-1", when we found workspace "w" that should be on "OUTPUT-2", we do the required change but then we continue to iterate over "OUTPUT-2".
We iterate again over the workspases of "OUTPUT-2".
END.
In the end only the first found workspace is correctly assigned to the display.

However I don't know how to fix this in a clean way because I didn't find a way to iterate over list while being modified.
